File: Connect4_student_pack/python/board.py
from typing import Tuple, List

import numpy as np


class Board:
    def __init__(self, board_size: Tuple[int, int], n_to_connect: int):
        self.__state = np.zeros(board_size, dtype=int)
        self.__board_size = board_size
        self.__n_to_connect = n_to_connect

        self.__n_steps = 0
        self.__last_player_index = None
        self.__last_player_row = None
        self.__last_player_column = None
        self.__winner = None

    def copy(self):
        new_board = Board(self.__board_size, self.__n_to_connect)
        new_board.__state = self.__state.copy()
        new_board.__board_size = self.__board_size
        new_board.__n_to_connect = self.__n_to_connect
        new_board.__n_steps = self.__n_steps
        new_board.__last_player_index = self.__last_player_index
        new_board.__last_player_row = self.__last_player_row
        new_board.__last_player_column = self.__last_player_column
        new_board.__winner = self.__winner

        return new_board

    def get_state(self) -> np.ndarray:
        return self.__state

    def step_is_valid(self, column: int):
        valid_col_index = 0 <= column < self.__board_size[1]
        if not valid_col_index:
            return False
        is_space_on_the_top = self.__state[0][column] == 0
        return is_space_on_the_top

    def get_valid_steps(self) -> List[int]:
        valid_cols = []
        col = 0
        while col < self.__board_size[1]:
            if self.step_is_valid(col):
                valid_cols.append(col)
            col += 1
        return valid_cols

    def step(self, player_index: int, column: int) -> None:
        self.__n_steps += 1
        self.__last_player_index = player_index
        self.__last_player_column = column
        row = self.__board_size[0] - 1
        while row >= 0:
            if self.__state[row][column] == 0:
                self.__state[row][column] = player_index
                self.__last_player_row = row
                return
            row -= 1

    def game_ended(self) -> bool:
        # board is unused
        if self.__last_player_index is None:
            return False
        # player won
        last_player_won = (
            self.is_n_in_a_row(self.__last_player_row, self.__last_player_column, self.__last_player_index) or
            self.is_n_in_a_col(self.__last_player_row, self.__last_player_column, self.__last_player_index) or
            self.is_n_diagonally(self.__last_player_row, self.__last_player_column, self.__last_player_index) or
            self.is_n_skew_diagonally(self.__last_player_row, self.__last_player_column, self.__last_player_index))
        if last_player_won:
            self.__winner = self.__last_player_index
            return True
        # table is full
        table_is_full = self.__n_steps == self.__board_size[0] * self.__board_size[1]
        if table_is_full:
            self.__winner = 0
            return True
        return False

    def is_n_in_a_row(self, row: int, col: int, player_index: int) -> bool:
        n_in_a_row = 0
        start_col = max(0, col - self.__n_to_connect + 1)
        end_col = min(self.__board_size[1], col + self.__n_to_connect)
        c = start_col
        while c < end_col:
            if self.__state[row][c] == player_index:
                n_in_a_row += 1
                if n_in_a_row == self.__n_to_connect:
                    self.__winner = player_index
                    return True
            else:
                n_in_a_row = 0
            c += 1
        return False

    def is_n_in_a_col(self, row: int, col: int, player_index: int) -> bool:
        n_in_a_col = 0
        start_row = max(0, row - self.__n_to_connect + 1)
        end_row = min(self.__board_size[0], row + self.__n_to_connect)
        r = start_row
        while r < end_row:
            if self.__state[r][col] == player_index:
                n_in_a_col += 1
                if n_in_a_col == self.__n_to_connect:
                    self.__winner = player_index
                    return True
            else:
                n_in_a_col = 0
            r += 1
        return False

    def is_n_diagonally(self, row: int, col: int, player_index: int) -> bool:
        n_in_a_diagonal = 0
        step_left_up = min(self.__n_to_connect - 1, min(row, col))
        step_right_down = min(self.__n_to_connect, min(self.__board_size[0] - row, self.__board_size[1] - col))
        if (step_left_up + step_right_down) < self.__n_to_connect:
            return False
        diagonal_step = -step_left_up
        while diagonal_step < step_right_down:
            if self.__state[row + diagonal_step][col + diagonal_step] == player_index:
                n_in_a_diagonal += 1
                if n_in_a_diagonal == self.__n_to_connect:
                    self.__winner = player_index
                    return True
            else:
                n_in_a_diagonal = 0
            diagonal_step += 1
        return False

    def is_n_skew_diagonally(self, row: int, col: int, player_index: int) -> bool:
        n_in_a_skew_diagonal = 0
        step_left_down = min(self.__n_to_connect - 1, min(self.__board_size[0] - row - 1, col))
        step_right_up = min(self.__n_to_connect, min(row + 1, self.__board_size[1] - col))
        if (step_right_up + step_left_down) < self.__n_to_connect:
            return False
        skew_diagonal_step = -step_left_down
        while skew_diagonal_step < step_right_up:
            if self.__state[row - skew_diagonal_step][col + skew_diagonal_step] == player_index:
                n_in_a_skew_diagonal += 1
                if n_in_a_skew_diagonal == self.__n_to_connect:
                    self.__winner = player_index
                    return True
            else:
                n_in_a_skew_diagonal = 0
            skew_diagonal_step += 1
        return False

    def get_winner(self) -> int:
        return self.__winner

    def get_last_player_index(self) -> int:
        return self.__last_player_index

    def get_last_player_row(self) -> int:
        return self.__last_player_row

    def get_last_player_column(self) -> int:
        return self.__last_player_column

File: Connect4_student_pack/src/Board.java
import java.util.ArrayList;
import java.util.Arrays;

import static java.lang.Math.max;
import static java.lang.Math.min;

public class Board {
    private final int[][] state;
    private final int[] boardSize;
    private final int nToConnect;
    private int nSteps = 0;
    private int lastPlayerIndex = -1;

    private int lastPlayerRow = -1;
    private int lastPlayerColumn = -1;
    private int winner = -1;

    public Board(int[] boardSize, int nToConnect) {
        this.state = new int[boardSize[0]][boardSize[1]];
        this.boardSize = boardSize;
        this.nToConnect = nToConnect;
    }

    public Board(Board boardToCopy) {
        this.state = Arrays.stream(boardToCopy.state).map(int[]::clone).toArray(int[][]::new);
        this.boardSize = boardToCopy.boardSize;
        this.nToConnect = boardToCopy.nToConnect;
        this.nSteps = boardToCopy.nSteps;
        this.lastPlayerIndex = boardToCopy.lastPlayerIndex;
        this.lastPlayerRow = boardToCopy.lastPlayerRow;
        this.lastPlayerColumn = boardToCopy.lastPlayerColumn;
        this.winner = boardToCopy.winner;
    }

    public int[][] getState() {
        return state;
    }

    public boolean stepIsValid(int column) {
        boolean validColIndex = column >= 0 && column < this.boardSize[1];
        if (!validColIndex)
            return false;

        boolean isSpaceOnTheTop = this.state[0][column] == 0;

        return isSpaceOnTheTop;
    }

    public ArrayList<Integer> getValidSteps() {
        ArrayList<Integer> validCols = new ArrayList<Integer>();
        for (int col = 0; col < boardSize[1]; col++) {
            if (stepIsValid(col)) {
                validCols.add(col);
            }
        }

        return validCols;
    }

    public void step(int playerIndex, int column) {
        nSteps++;
        lastPlayerIndex = playerIndex;
        lastPlayerColumn = column;
        for (int row = boardSize[0] - 1; row >= 0; row--) {
            if (this.state[row][column] == 0) {
                this.state[row][column] = playerIndex;
                lastPlayerRow = row;
                return;
            }
        }
    }

    public boolean gameEnded() {
        //board is unused
        if (lastPlayerIndex == -1) {
            return false;
        }

        // player won
        boolean lastPlayerWon =
                isNInARow(lastPlayerRow, lastPlayerColumn, lastPlayerIndex) ||
                        isNInACol(lastPlayerRow, lastPlayerColumn, lastPlayerIndex) ||
                        isNDiagonally(lastPlayerRow, lastPlayerColumn, lastPlayerIndex) ||
                        isNSkewDiagonally(lastPlayerRow, lastPlayerColumn, lastPlayerIndex);

        if (lastPlayerWon) {
            winner = lastPlayerIndex;
            return true;
        }

        // table is full
        boolean tableIsFull = nSteps == boardSize[0] * boardSize[1];
        if (tableIsFull) {
            winner = 0;
            return true;
        }

        return false;
    }

    private boolean isNInARow(int row, int col, int playerIndex) {
        int nInARow = 0;

        int startCol = max(0, col - nToConnect + 1);
        int endCol = min(boardSize[1], col + nToConnect);

        for (int c = startCol; c < endCol; c++) {
            if (state[row][c] == playerIndex) {
                nInARow++;
                if (nInARow == nToConnect) {
                    winner = playerIndex;
                    return true;
                }
            } else
                nInARow = 0;
        }
        return false;
    }

    private boolean isNInACol(int row, int col, int playerIndex) {
        int nInACol = 0;

        int startRow = max(0, row - nToConnect + 1);
        int endRow = min(boardSize[0], row + nToConnect);

        for (int r = startRow; r < endRow; r++) {
            if (state[r][col] == playerIndex) {
                nInACol++;
                if (nInACol == nToConnect) {
                    winner = playerIndex;
                    return true;
                }
            } else
                nInACol = 0;
        }
        return false;
    }

    private boolean isNDiagonally(int row, int col, int playerIndex) {
        int nInADiagonal = 0;

        int stepLeftUp = min(nToConnect - 1, min(row, col));
        int stepRightDown = min(nToConnect, min(boardSize[0] - row, boardSize[1] - col));

        if ((stepLeftUp + stepRightDown) < nToConnect)
            return false;

        for (int diagonalStep = -stepLeftUp; diagonalStep < stepRightDown; diagonalStep++) {
            if (state[row + diagonalStep][col + diagonalStep] == playerIndex) {
                nInADiagonal++;
                if (nInADiagonal == nToConnect) {
                    winner = playerIndex;
                    return true;
                }
            } else {
                nInADiagonal = 0;
            }
        }
        return false;
    }

    private boolean isNSkewDiagonally(int row, int col, int playerIndex) {
        int nInASkewDiagonal = 0;

        int stepLeftDown = min(nToConnect - 1, min(boardSize[0] - row - 1, col));
        int stepRightUp = min(nToConnect, min(row + 1, boardSize[1] - col));

        if ((stepRightUp + stepLeftDown) < nToConnect)
            return false;

        for (int skewDiagonalStep = -stepLeftDown; skewDiagonalStep < stepRightUp; skewDiagonalStep++) {
            if (state[row - skewDiagonalStep][col + skewDiagonalStep] == playerIndex) {
                nInASkewDiagonal++;
                if (nInASkewDiagonal == nToConnect) {
                    winner = playerIndex;
                    return true;
                }
            } else
                nInASkewDiagonal = 0;
        }
        return false;
    }

    public int getWinner() {
        return winner;
    }

    public int getLastPlayerIndex() {
        return lastPlayerIndex;
    }

    public int getLastPlayerRow() {
        return lastPlayerRow;
    }

    public int getLastPlayerColumn() {
        return lastPlayerColumn;
    }
}
